### PR TITLE
Add `SRA`, `MULHS`, `MULHU` instructions

### LINKED ISSUE #13: Implement `MULHU` and `NE` instructions
These instructions are needed for the MVP. We need `MULHU` for immediate division and `NE` for the logical operators (and, or, not). Following the format of our [ISA](https://github.com/valida-xyz/valida-compiler/issues/2), the spec for these instructions are:

`NE`:
Operands: a, b(fp), c(fp)
Description: If [b] != [c] then set a to 1. Otherwise, set a to 0. 

`MULHU`:
Operands: a, b(fp), c(fp)
Description: Multiply [b] by [c] then set a to the top part of the result.

### PR DESCRIPTION
**Pull request by thealmarty**
Closes #13. (`NE` to be done in the compiler.) Also I removed some unneeded code that I added in PR #16. Again the STARKs are not done. See #17.

Before these changes, the following tests ran with errors :
2023-12-divi, 2023-12-divi-neg, 2023-12-divi-signed, 2023-12-divi-signed-neg

After these changes, the tests ran with the expected results (positive tests don't terminate, `neg` tests terminate). Steps to reproduce (make sure to pull `llvm-test-suite`):
```
$ cd
$ mkdir -p code/lita/issue/13
$ cd code/lita/issue/13
$ git clone --depth 1 --branch develop git@github.com:lita-xyz/llvm-valida.git
$ git clone --depth 1 --branch develop git@github.com:lita-xyz/llvm-test-suite.git
$ git clone git@github.com:lita-xyz/valida.git
$ git clone git@github.com:lita-xyz/valida-mc-fixer.git
$ git clone git@github.com:lita-xyz/Plonky3.git
$ cmake -G Ninja -DCMAKE_BUILD_TYPE=Debug -DLLVM_TARGETS_TO_BUILD=X86 -DLLVM_EXPERIMENTAL_TARGETS_TO_BUILD=Delendum -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DLLVM_USE_LINKER=lld -DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra;libc;lld" -B llvm-valida/build -S llvm-valida/llvm
$ cmake --build llvm-valida/build --target clang -j7 # substitute for 7 the number of cores to use to build
$ cmake --build llvm-valida/build --target lld -j7
$ cmake --build llvm-valida/build --target llvm-objcopy -j7
$ nix build ./valida-mc-fixer -o ./valida-mc-fixer/result
$ cd valida
$ cargo build --release
$ cd ..
$ ./llvm-valida/build/bin/clang -c -target delendum ./llvm-valida/DelendumEntryPoint.c -o ./llvm-valida/build/DelendumEntryPoint.o
# replace $test with the actual test
$ ./llvm-valida/build/bin/clang -c -target delendum ./llvm-test-suite/SingleSource/UnitTests/$test.c -o $test.o
$ ./llvm-valida/build/bin/ld.lld --script=./llvm-valida/valida.ld -o ./$test.out ./$test.o 
$ ./llvm-valida/build/bin/llvm-objcopy -O binary ./$test.out ./$test.out.bin
$ ./valida-mc-fixer/result/bin/valida-mc-fixer ./$test.out.bin ./$test.out.bin.fix
$ ./valida/target/release/valida ./$test.out.bin.fix
```

### CODE REVIEW

File: alu_u32/src/shift/mod.rs [hunk 1]
```diff
@@ -241,7 +248,52 @@ where
         state
             .shift_u32_mut()
             .operations
-            .push(Operation::Shl32(a, b, c));
```
Comment (thealmarty) [inline alu_u32/src/shift/mod.rs:244]: I think this was a mistake.

File: alu_u32/src/shift/mod.rs [hunk 2]
```diff
@@ -119,11 +122,15 @@ impl Shift32Chip {
...
-                cols.is_shl = F::one();
+                cols.is_shr = F::one();
+                self.set_cols(cols, a, b, c);
+            }
+            Operation::Sra32(a, b, c) => {
+                cols.is_sra = F::one();
                 self.set_cols(cols, a, b, c);
             }
             Operation::Shl32(a, b, c) => {
-                cols.is_shr = F::one();
```
Comment (thealmarty) [inline alu_u32/src/shift/mod.rs:126]: I think this was a mistake.

File: alu_u32/src/shift/mod.rs [hunk 3]
```diff
@@ -119,11 +122,15 @@ impl Shift32Chip {
 
         match op {
             Operation::Shr32(a, b, c) => {
-                cols.is_shl = F::one();
```
Comment (thealmarty) [inline alu_u32/src/shift/mod.rs:122]: I think this was a mistake.

**Comment by thealmarty** [comment]
I've also add 2023-12-sra, 2023-12-sra-neg tests and they are working as expected. Since `c` doesn't have `mulh` natively, no new tests are added for them (the `divi` tests are already indirectly testing them).

**Comment by morganthomas** [comment]
Hi @thealmarty ! I have reproduced your results. I created a quick shell script which runs your tests. I am sharing this with you in case you find it helpful and to suggest that in the future you present your test scripts in such a format, using bash's facilities for looping and substitution.

```bash
morgan@aristotle:~/code/lita/issue/13$ cat run-tests.sh 
#!/usr/bin/env bash
# set -x   # uncomment to see commands that are run printed
TESTS="2023-12-divi 2023-12-divi-neg 2023-12-divi-signed 2023-12-divi-signed-neg"
for test in $TESTS
do
	echo $test
	./llvm-valida/build/bin/clang -c -target delendum ./llvm-test-suite/SingleSource/UnitTests/$test.c -o $test.o
	./llvm-valida/build/bin/ld.lld --script=./llvm-valida/valida.ld -o ./$test.out ./$test.o
	./llvm-valida/build/bin/llvm-objcopy -O binary ./$test.out ./$test.out.bin
	./valida-mc-fixer/result/bin/valida-mc-fixer ./$test.out.bin ./$test.out.bin.fix
	timeout 3 ./valida/target/release/valida ./$test.out.bin.fix
	echo $? # status 124 indicates non-termination; status 0 indicates successful termination
done
morgan@aristotle:~/code/lita/issue/13$ ./run-tests.sh 
2023-12-divi
124
2023-12-divi-neg
0
2023-12-divi-signed
124
2023-12-divi-signed-neg
0
```

Note that you need to `chmod +x ./run-tests.sh` or the equivalent, in order to make the script executable, if you want to run it simply by typing its path in the shell, as illustrated above.

**Review by morganthomas — APPROVED** [review]

**Pull request merged**